pkgs/fc/ceph/src/fc/ceph/lvm.py

```diff
@@ -10,7 +10,92 @@
 from fc.ceph.util import console, mount_status, run
 
 
+class GenericBlockDevice:
+    def __new__(cls, name: str):
+        # prevent explicitly instantiated child classes from returning as None
+        if cls is not GenericLogicalVolume:
+            return object.__new__(cls)
+        if MdraidDevice.exists(name):
+            return MdraidDevice(name)
+        elif PartitionedDisk.exists(name):
+            return PartitionedDisk(name)
+        # Allow operations to proceed by creating a block device
+        return None
+
+    def __init__(self, name: str):
+        self.name = name
+
+    # create cannot be a generic method, as it requires different params depending on the type
+    @property
+    def blockdevice(self) -> str:
+        raise NotImplementedError
+
+    @classmethod
+    def exists(cls, name: str) -> bool:
+        raise NotImplementedError
+
+
+class MdraidDevice(GenericBlockDevice):
+    """represents a software RAID array in mode RAID6 with spare disk. Uses the
+    whole provided blockdevices without partitioning."""
+
+    @classmethod
+    def create(cls, name: str, blockdevices: list[str]):
+        obj = cls(name)
+
+    def __init__(self, name: str):
+        self.name = name
+
+    @property
+    def blockdevice(self) -> str:
+        return f"/dev/md/{self.name}"
+
+    @classmethod
+    def exists(cls, name) -> bool:
+        return name.startswith("/dev/md") and os.path.exists(f"/dev/md/{name}")
+
+    # TODO do we need activating?
+
+
+class PartitionedDisk(GenericBlockDevice):
+    """name denotes the path to the whole unpartitioned disk"""
+
+    def __init__(self, name: str):
+        super().__init__(name)
+        self._blockdevice: Optional[str] = None
+
+    @staticmethod
+    def _partition_candidates(disk: str) -> list[str]:
+        return [f"{disk}1", f"{disk}p1"]
+
+    @property
+    def blockdevice(self):
+        if not self._blockdevice:
+            for partition in self._partition_candidates(self.name):
+                if os.path.exists(partition):
+                    self._blockdevice = partition
+                    break
+            else:
+                raise RuntimeError(f"Could not find partition 1 on {self.name}")
+        return self._blockdevice
+
+    @classmethod
+    def create(cls, disk: str) -> "PartitionedDisk":
+        obj = cls(disk)
+        run.sgdisk("-Z", disk)
+        run.sgdisk("-a", "8192", "-n", "1:0:0", "-t", "1:8e00", disk)
```

Comment:
`8e00` is the LVM type, this needs to change depending on which layer we put on top of this. Could e.g. be an mdraid instead.

Provide a `ensure_partition_type` method for classes using such a device.